pkg/observable/channel/replay.go

```diff
@@ -0,0 +1,209 @@
+package channel
+
+import (
+	"context"
+	"log"
+	"sync"
+	"time"
+
+	"pocket/pkg/observable"
+)
+
+// TODO_CONSIDERATION: perhaps this should be parameterized.
+
+// replayPartialBufferTimeout is the duration to wait for the replay buffer to
+// accumulate at least 1 value before returning the accumulated values.
+const replayPartialBufferTimeout = 100 * time.Millisecond
+
+var _ observable.ReplayObservable[any] = (*replayObservable[any])(nil)
+
+type replayObservable[V any] struct {
+	// observableInternals[V] is embedded to provide safe read and write access
+	// to the observers list.
+	observableInternals[V]
+	// replayBufferSize is the number of notifications to buffer so that they
+	// can be replayed to new observers.
+	replayBufferSize int
+	// replayBufferMu protects replayBuffer from concurrent access/updates.
+	replayBufferMu sync.RWMutex
+	// replayBuffer holds the last relayBufferSize number of notifications received
+	// by this observable. This buffer is replayed to new observers, on subscribing,
+	// prior to any new notifications being propagated.
+	replayBuffer []V
+}
+
+// NewReplayObservable returns a new ReplayObservable with the given replay buffer
+// replayBufferSize and the corresponding publish channel to notify it of new values.
+func NewReplayObservable[V any](
+	ctx context.Context,
+	replayBufferSize int,
+) (observable.ReplayObservable[V], chan<- V) {
+	obsvbl, publishCh := NewObservable[V]()
+	return ToReplayObservable[V](ctx, replayBufferSize, obsvbl), publishCh
+}
+
+// ToReplayObservable returns an observable which replays the last replayBufferSize
+// number of values published to the source observable to new observers, before
+// publishing new values.
+// It panics if srcObservable does not implement the observableInternals interface.
```

Comment:
If we end up renaming `observableInternals`, make sure to update the comments too.

Reply:
:pray: